### PR TITLE
Feat(Seg): left-over changes

### PR DESCRIPTION
**Pull request by jdeschamps**
## Description

> [!IMPORTANT]  
> This PR contains the remnants of the changes that were not split into the multiple sub-PRs:
> - #1061
> - #1062
> - #1067
> - #1068
> - #1081

Introduces the backbone for semantic segmentation in CAREamics. While we are predominantly a restoration library, semantic segmentation in a codebase that we know inside out is valuable for the facility and potentially our users. In addition CAREamics has strong arguments with the zarr support, and we could benefit from that in segmentation projects.

### Input format

`targets` should be passed as class labels and are assumed to have same `axis` as the inputs, albeit without a `C` dimension.

### Internal change to model constraints

The main departure from the previous behaviour is that in a segmentation task the number of model output channels is different from the targets. The former are probabilities with `N` channels, while the `target` has class labels (singleton channel after reshaping). In order to pass the model constraints, I needed to add `model_constraints` that are algorithm-aware, meaning that `get_model_constraints` now take the `algorithm_config` rather than the `model_config`.

```python
# before
def get_model_constraints(model_config: UNetConfig) -> ModelConstraints:

# now
def get_model_constraints(algorithm_config: AlgorithmConfig) -> ModelConstraints:
```

### Bulk of changes

The bulk of changes pertains to making all the necessary segmentation classes and modules:
- Segmentation configuration
- Segmentation algorithm configuration
- Segmentation configuration factory
- Segmentation Lightning module
- Segmentation losses

### Out-of-scope changes

Note that this PR also adds the following out-of-scope features: 
- Adding missing `target_axes` to the prediction functions in `CAREamist`. 
- Added an example folder (`/examples/`), from which `.gitignore` ignores all files but `.md`, `.py` and `.ipynb`
- e2e CI running only smoke/e2e tests from tests/e2e
- Fixed pre-commit, `int` casting in `xy_flip.py`

⚠️ This PR includes #1058 and should therefore be reviewed after merging it.

### API

Nothing changed there:

```python
    cfg = create_seg_config(
        experiment_name="test_seg",
        data_type="array",
        axes="SYX",
        batch_size=4,
        patch_size=(64, 64),
        num_epochs=20,
        n_classes=n_classes,
    )

    careamist = CAREamist(config=cfg, work_dir=tmp_path)
    careamist.train(
        train_data=train_data,
        train_data_target=train_data_tar,
        val_data=val_data,
        val_data_target=val_data_tar,
    )

    # predict to memory
    pred, _ = careamist.predict(pred_data=train_data)
```

## Split in misc PRs

- [x] pre-commit error #1060
- [x] e2e #1065
- [x] model constraints #1061
- [x] `target_axes` #1063
- [x] losses #1062 
- [x] configurations: #1067
- [x] lightning module #1068

### Added

- Added `SegConfiguration`
    - Validates that `skip_target` is `True`
- Added `SegAlgorithm` configuration
    - Three losses are available: dice, cross-entropy and dice+cross-entropy
    - The U-Net should have at least 2 channels (background and foreground for a binary segmentation)
- Added segmentation configuration factories
    - No `target_axes`, as it is inferred from `axes` (same as `axes`, without "C")
    - Parameter `n_classes` which are the number of **foreground** classes (1 for binary, >1 for multiclass), background class is automatically added.
    - Normalization is enforced with `skip_target==True`
- Added segmentation lightning module `SegModule`
    - Losses are in charge of turning logits into probability/class indices
    - Prediction currently does not output prediction but the argmax. 
- Added segmentation losses
    - Dice, Cross-entropy, and Dice+Cross-entropy
- Added segmentation constraints
    - Inherits from `UNetConstraints` to re-use the spatial and input channels constraints
    - Since the maximum label from a class label target cannot be checked (e.g. a patch may not contain all labels), the constraints only check that the number of channels is equal to 1.
- Added tests, including e2e
- Added `examples/segmentation` notebooks

### Modified
- Algorithm configs:
    - `UNetAlgorithmConfig` now takes a general `str` for most of its parameters, to prevent the explosion of `Literal` choices
- Classes and modules extended to accommodate segmentation algorithm/losses:
    - algorithm factory
    - config discriminators
    - training configuration and factory
    - supported algorithm and losses
- Update model constraints `get_model_constraints` to take the algorithm config rather than the model configuration, in order to branch out depending on the algorithm. 
- Renamed `create_ng_data_configuration` to `create_data_configuration`, this is unfortunately also the case in another PR and led to many files being changed
- `CAREamist` has now `target_axes` parameter in the prediction function, something that was forgotten in related PRs.
- `ImageWriteStrtegy` now performs axis restoration to ensure that direct prediction and prediction to disk return the same array.
- Various reformulation of error raised and missing docstrings.

### Single-channel (Dice)
dice_single

### Multi-channel (DiceCE)
multi_dicece

### All metrics

| Task  | Loss | Class | IoU mu +/- std |
| ------------- | ------------- | ------------- | ------------- |
| Single-class  | CE  | Class 0  |0.98 +/- 0.03
| | | Class 1| 0.87 +/- 0.08
| Single-class  | Dice  | Class 0 | 0.98 +/- 0.03
| | | Class 1 | 0.87 +/- 0.08
| Single-class  | DiceCE  | Class 0| 0.97 +/- 0.03
| | | Class 1| 0.86 +/- 0.08
| Multi-class  | CE  | Class 0 | 0.98 +/- 0.03
| | | Class 1 | 0.80 +/- 0.14
| | | Class 2 | 0.63 +/- 0.15
| Multi-class  | Dice  | Class 0 | 0.97 +/- 0.03
| | | Class 1 | 0.79 +/- 0.14
| | | Class 2 | 0.63 +/- 0.15
| Multi-class  | DiceCE  | Class 0 | 0.98 +/- 0.03
| | | Class 1 |  0.80 +/- 0.14
| | | Class 2 | 0.63 +/- 0.16

## Notes

The decision here is that we expect targets to be given has a label image (pixel values encode classes), while the model outputs N (= number of classes) channels.

## Roadmap for integration

Current road map for semantic segmentation:
- [x] Skip target normalization
    - https://github.com/CAREamics/careamics/issues/680
    - [x] https://github.com/CAREamics/careamics/pull/985
- [x] Allow loading targets with different axes
    - https://github.com/CAREamics/careamics/issues/963
    - [x] https://github.com/CAREamics/careamics/pull/1005 
- [x] Allow prediction output to have different shape than input (e.g. channels)
    - https://github.com/CAREamics/careamics/issues/948
    - [x] https://github.com/CAREamics/careamics/pull/956
    - [x] https://github.com/CAREamics/careamics/pull/993
- [x] Segmentation model constraints to avoid error when comparing model output and target size
    - https://github.com/CAREamics/careamics/issues/964
    - [x] https://github.com/CAREamics/careamics/pull/986/changes/1d49422573a12264c54e6519c353a603db8c8055
-  [x] Multiclass metrics
    - [x] https://github.com/CAREamics/careamics/pull/986/changes/0de7c6b641de20879abd8387e216c0d4ee8b9f2d 
-  [x] Configuration and module
    - [x] https://github.com/CAREamics/careamics/pull/986/changes/e90a317a562931d5a69f29e26a2c08a62992f445 
- [x] Update config factory to reflect refactoring from #1005 
    - https://github.com/CAREamics/careamics/pull/986/changes/be07d05a8efc438ca13d698b268f01e5659bc4ea
- [x] Look at validation step in module
- [x] Check if model constraints on input channels can be re-used from UNet constraints
- [x] Enforce `skip_target` at the level of the configuration
- [ ] Refactor algorithm config to follow recent changes from BMZ PR
- [x] Review tests and e2e
- [x] Proof of principle, single and multi-class
- [x] Test with C and without
- [ ] Test with CZI
- [x] Benchmark
- [x] Review tests, move tests to `tests.unit`
- [ ] Docs
- [ ] Open issue to predict class probability

### CODE REVIEW

File: src/careamics/config/factories/seg_unet_factory.py [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+
+    If "Z" is present in `axes`, then `patch_size` must be a list of length 3, otherwise
+    2.
+
+    If "C" is present in `axes`, then you need to set `n_channels_in` to the number
+    of input channels.
+
+    By default, the transformations applied are a random flip along X or Y, and a random
+    90 degrees rotation in the XY plane. Normalization is always applied, as well as the
+    N2V manipulation.
```
Comment (jdeschamps) [inline src/careamics/config/factories/seg_unet_factory.py]:
```suggestion
    90 degrees rotation in the XY plane. Normalization is always applied.
```

File: src/careamics/config/factories/seg_unet_factory.py [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    if n_channels_in is None:
+        n_channels_in = 1 if channels is None else len(channels)
+
+    # normalization
+    norm_config = {"name": normalization, "skip_target": True}
+    if normalization_params is not None:
+        norm_config.update(normalization_params)
+
+    # augmentations
+    augs: list[XYFlipConfig | XYRandomRotate90Config] | None = None
```
Comment (jdeschamps) [inline src/careamics/config/factories/seg_unet_factory.py]:
```suggestion
    # TODO refactor after target_axes PR will be merged
    augs: list[XYFlipConfig | XYRandomRotate90Config] | None = None
```

File: src/careamics/losses/segmentation_losses.py [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            # classes to include
+            start_idx = 0 if self.include_background else 1
+            dice_per_class = dice_per_class[:, start_idx:]
+
+            # apply weights if provided
+            if self.weight is not None:
+                weight = self.weight[start_idx:].to(dice_per_class.device)
+                dice_per_class = dice_per_class * weight
+
+            # Average across classes and batch
```
Comment (jdeschamps) [inline src/careamics/losses/segmentation_losses.py]:
```suggestion
            # average across classes and batch
```
